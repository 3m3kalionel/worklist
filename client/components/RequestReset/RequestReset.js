import React from 'react';
import { Mutation } from 'react-apollo';

import AuthFormStyles from '../../styles/AuthFormStyles';
import WithForm from '../HOC';
import REQUEST_RESET_MUTATION from '../../graphql/mutations/RequestReset';

const WrappedComponent = props => {
  const { onChange, onSubmit, onError, values = { email: ''} } = props;
  const { email } = values;

  return (
    <Mutation
      mutation={REQUEST_RESET_MUTATION}
      variables={{ recipientEmail: email }}
    >
      {(requestResetPasswordLink, { data, error, loading, called }) => (
          <form
            onSubmit={onSubmit(requestResetPasswordLink, 'requestreset')}
            data-testid="form"
          >
            <AuthFormStyles>
              <div>
                <div id="authentication-form">
                  <div id="form-body">
                    {error && onError(error)}
                    {props.children}
                    <div id="success-message">
                      {
                        !error && !loading && called &&
                        data.requestResetPasswordLink.message
                      }
                    </div>
                    <h1 id="form-header">Forgot Password ?</h1>
                    <label htmlFor="email">Email</label>
                    <input
                      name="email"
                      type="email"
                      className="auth-form-input"
                      placeholder="Email"
                      value={email}
                      autoComplete="email"
                      onChange={onChange}
                      required
                    />
                    <div id="action-button-container">
                      <button
                        type="submit"
                        id="submit-button"
                        disabled={loading}
                      >
                        Send Reset Link
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </AuthFormStyles>
          </form>
        )
      }
    </Mutation>
  )
}

export default WithForm(WrappedComponent);
