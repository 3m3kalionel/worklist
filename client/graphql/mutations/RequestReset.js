import gql from 'graphql-tag';

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($recipientEmail: String!) {
    requestResetPasswordLink(recipientEmail: $recipientEmail) {
      message
    }
  }
`;

export default REQUEST_RESET_MUTATION;
