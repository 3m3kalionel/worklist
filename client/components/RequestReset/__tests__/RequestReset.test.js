import React from 'react';
import {
  render,
  fireEvent,
} from '@testing-library/react';

import { ApolloProvider } from 'react-apollo';
import '@testing-library/react/cleanup-after-each';
import { createMockClient } from 'mock-apollo-client';

import RequestReset from '../RequestReset';

import REQUEST_RESET_MUTATION from '../../../graphql/mutations/RequestReset';

jest.mock("../../HOC.js", () => Component => {
  const mockHOC = props => (
    <div><Component {...props} /></div>
  );

   return mockHOC;
});

const setup = (values, onSubmit) => {
  const onSubmitHandler = onSubmit || jest.fn();
  const mockClient = createMockClient();

  mockClient.setRequestHandler(
    REQUEST_RESET_MUTATION,
    () => Promise.resolve({
      data: {
        requestResetPasswordLink: {
          message: 'Check your email for a reset link',
        }
      }
    })
  );

  const {
    container,
    getByText,
    getByTestId
  } = render(
    <ApolloProvider client={mockClient}>
      <RequestReset
        values={values}
        onSubmit={onSubmitHandler}
        onChange={jest.fn()}
      />
    </ApolloProvider>
  );

  return { container, getByText, getByTestId }
}

describe('<RequestReset />', () => {
  test('renders an empty form', async () => {
    const { container } = setup({ email: "" });

    expect(container).toMatchSnapshot();    
  });

  test('renders text in the form input', () => {
    const { container } = setup({ email: "email@gmail.com" });

    expect(container).toMatchSnapshot();
  });

  test('calls the submit handler when button is clicked', async () => {
    const onSubmitHandler = jest.fn();
    const {
      container,
      getByTestId
    } = setup({ email: "email@gmail.com" }, onSubmitHandler);
    const form = getByTestId('form');
    fireEvent.submit(form);

    expect(onSubmitHandler).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
})
