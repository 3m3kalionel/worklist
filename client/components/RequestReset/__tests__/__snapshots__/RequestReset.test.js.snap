// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset /> calls the submit handler when button is clicked 1`] = `
<div>
  <div>
    <form
      data-testid="form"
    >
      <div
        class="AuthFormStyles-js87jd-0 ieHHDD"
      >
        <div>
          <div
            id="authentication-form"
          >
            <div
              id="form-body"
            >
              <div
                id="success-message"
              />
              <h1
                id="form-header"
              >
                Forgot Password ?
              </h1>
              <label
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="auth-form-input"
                name="email"
                placeholder="Email"
                required=""
                type="email"
                value="email@gmail.com"
              />
              <div
                id="action-button-container"
              >
                <button
                  id="submit-button"
                  type="submit"
                >
                  Send Reset Link
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`<RequestReset /> renders an empty form 1`] = `
<div>
  <div>
    <form
      data-testid="form"
    >
      <div
        class="AuthFormStyles-js87jd-0 ieHHDD"
      >
        <div>
          <div
            id="authentication-form"
          >
            <div
              id="form-body"
            >
              <div
                id="success-message"
              />
              <h1
                id="form-header"
              >
                Forgot Password ?
              </h1>
              <label
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="auth-form-input"
                name="email"
                placeholder="Email"
                required=""
                type="email"
                value=""
              />
              <div
                id="action-button-container"
              >
                <button
                  id="submit-button"
                  type="submit"
                >
                  Send Reset Link
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`<RequestReset /> renders text in the form input 1`] = `
<div>
  <div>
    <form
      data-testid="form"
    >
      <div
        class="AuthFormStyles-js87jd-0 ieHHDD"
      >
        <div>
          <div
            id="authentication-form"
          >
            <div
              id="form-body"
            >
              <div
                id="success-message"
              />
              <h1
                id="form-header"
              >
                Forgot Password ?
              </h1>
              <label
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="auth-form-input"
                name="email"
                placeholder="Email"
                required=""
                type="email"
                value="email@gmail.com"
              />
              <div
                id="action-button-container"
              >
                <button
                  id="submit-button"
                  type="submit"
                >
                  Send Reset Link
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
